### ISSUE TITLE
[Question] BN and Dropout

### ISSUE THREAD

**Issue by YRFT** [ageron/handson-ml2#423]
In chpter 11 -> section "Avoiding Overfitting Through Regularization" -> "Monte Carlo (MC) Dropout". The book said:

> If your model contains other layers that behave in a special way during training (such as `BatchNormalization` layers), then you should ...

So, my question is when we use both batch normalization and dropout in a DNN. What should we use first?

I searched for relevant questions (e.g.  [Ordering of batch normalization and dropout?](https://stackoverflow.com/questions/39691902/ordering-of-batch-normalization-and-dropout)), but could not find a satisfactory answer.

Please give me some guidelines (maybe it just depends on the dataset?), thank you. I love this book badly.

---
Update:

There is [a paper](https://arxiv.org/pdf/1801.05134.pdf) which gives a few advices. Any suggestions?

**Comment by ageron** [comment]
Hi @YRFT ,

Thanks for your question and for your very kind words! 👍 

Understanding why BN and Dropout work well is still the subject of active research, and their interactions are not well understood, which is probably why you have some difficulty finding a clear answer.

The paper you point to seems to suggest one reason why BN and Dropout often don't cooperate very well. That said, many SOTA (State-Of-The-Art) neural net architectures use a combination of both, although not at every layer. For example, in SOTA CNNs, there's generally an abundance of BN layers, typically one per convolutional layer, but they typically don't have any Dropout layer attached to them. When there are a few Dense layers on top of these convolutional layers, the Dense layers often have a Dropout layer attached to them, but no BN layer. So the type of regularization layer that works best depends on the type of layers you are using, and on other factors like the neural net architecture and the dataset.

In short... it depends, and for now it's still quite empirical. Perhaps the best option for the moment is to check out the successful neural net architectures and get inspiration from them.

Hope this helps.

**Comment by YRFT** [comment]
Thank you for your thoughtful answer. It's very valuable to me, as this book.

ML is changing fast, we need a master to give us a leading :smile:. I am looking forward to the third edition.
